src/engine/strat_engine/filesystem.rs

```diff
@@ -47,11 +56,35 @@ impl StratFilesystem {
         }
     }
 
-    pub fn check(&self, dm: &DM) -> EngineResult<FilesystemStatus> {
+    /// check if filesystem is getting full and needs to be extended
+    /// TODO: deal with the thindev in a Fail state.
+    pub fn check(&mut self, dm: &DM) -> EngineResult<FilesystemStatus> {
         match try!(self.thin_dev.status(dm)) {
-            ThinStatus::Good((_mapped, _highest)) => {
-                // TODO: check if filesystem is getting full and might need to
-                // be extended (hint: use statfs(2))
+            ThinStatus::Good((_, _)) => {
+                let mount_point = try!(self.get_mount_point());
+                let (fs_total_bytes, fs_total_used_bytes, fs_block_size) =
+                    try!(fs_usage(&mount_point));
+                let free_bytes = Bytes((*fs_total_bytes - *fs_total_used_bytes) / *fs_block_size);
```

Comment:
If you changed ```fs_usage``` as suggested, this calculation becomes completely unnecessary, because ```fs_usage``` has returned the value you actually need.